src/xmipp/libraries/reconstruction_starpu/reconstruct_fourier_starpu_util.h

```diff
@@ -0,0 +1,430 @@
+			for (int x = 0; x < xSize; x++) {
+				where[z][y][x] = (T) 0;
+			}
+		}
+	}
+	return where;
+}
+
+/** Method to release 3D array of given size */
+template<typename T>
+static void release(T***& array, int ySize, int zSize) {
+	for(int z = 0; z < zSize; z++) {
+		for(int y = 0; y < ySize; y++) {
+			delete[] array[z][y];
+		}
+		delete[] array[z];
+	}
+	delete[] array;
+	array = NULL;
+}
+
+/** Method will copy continuous 3D arrays (with each side of `size`) to non-continuous 3D arrays */
+template<typename T>
+void copyFlatTo3D(T*** target, const T* source, const int size) {
+	for (int z = 0; z < size; z++) {
+		for (int y = 0; y < size; y++) {
+			int index = (z * size * size) + (y * size);
+			memcpy(target[z][y], source + index, size * sizeof(T));
+		}
+	}
+}
+
+
+/** Function behaving like an identity, i.e returning passed value */
+template<typename T>
+static T identity(T val) { return val;}; // if used with some big type, use reference
+
+/** Function returning conjugate of a complex number */
+template<typename T>
+static std::complex<T> conjugate(std::complex<T> f) { return conj(f);};
+
+
+/** Do 3x3 x 1x3 matrix-vector multiplication */
+static inline void multiply(const float transform[3][3], Point3D<float>& inOut) {
+	float tmp0 = transform[0][0] * inOut.x + transform[0][1] * inOut.y + transform[0][2] * inOut.z;
+	float tmp1 = transform[1][0] * inOut.x + transform[1][1] * inOut.y + transform[1][2] * inOut.z;
+	float tmp2 = transform[2][0] * inOut.x + transform[2][1] * inOut.y + transform[2][2] * inOut.z;
+	inOut.x = tmp0;
+	inOut.y = tmp1;
+	inOut.z = tmp2;
+}
+
+/**
+*          6____5
+*         2/___1/
+*    +    | |  ||   y
+* [0,0,0] |*|7 ||4
+*    -    |/___|/  z  sizes are padded with blob-radius
+*         3  x  0
+* [0,0] is in the middle of the left side (point [2] and [3]), provided the blobSize is 0
+* otherwise the values can go to negative values
+* origin(point[0]) in in 'high' frequencies so that possible numerical instabilities are moved to high frequencies
+*/
+static void createProjectionCuboid(Point3D<float> (&cuboid)[8], float sizeX, float sizeY, float blobSize) {
+	float halfY = sizeY / 2.0f;
+	cuboid[3].x = cuboid[2].x = cuboid[7].x = cuboid[6].x = 0.f - blobSize;
+	cuboid[0].x = cuboid[1].x = cuboid[4].x = cuboid[5].x = sizeX + blobSize;
+
+	cuboid[3].y = cuboid[0].y = cuboid[7].y = cuboid[4].y = -(halfY + blobSize);
+	cuboid[1].y = cuboid[2].y = cuboid[5].y = cuboid[6].y = halfY + blobSize;
+
+	cuboid[3].z = cuboid[0].z = cuboid[1].z = cuboid[2].z = 0.f + blobSize;
+	cuboid[7].z = cuboid[4].z = cuboid[5].z = cuboid[6].z = 0.f - blobSize;
+}
+
+/** Apply rotation transform to cuboid */
+static inline void rotateCuboid(Point3D<float> (&cuboid)[8], const float transform[3][3]) {
+	for (int i = 0; i < 8; i++) {
+		multiply(transform, cuboid[i]);
+	}
+}
+
+/** Add 'vector' to each element of 'cuboid' */
+static inline void translateCuboid(Point3D<float> (&cuboid)[8], Point3D<float> vector) {
```

Comment:
can you use original functions to avoid duplication of the code?

Reply:
I probably can, but I don't want to pull in more code than what is necessary, so I'll see what I can do. Additionally, if the functions are used in multiple programs, they should probably be moved to a common neutral folder, to prevent any surprises when it inevitably gets modified later.

Reply:
Indeed. Unfortunately, the folder housekeeping won't happen any time soon.